NPE on textDocument/definition on unbalanced section

Given this qute template, where the `#if` section is not closed
```
{#if foo}
```

if you ctrl+hover over `if`, `textDocument/definition` throws an NPE:
```
SEVERE: Internal error: java.lang.NullPointerException: Property must not be null: targetRange
java.util.concurrent.CompletionException: java.lang.NullPointerException: Property must not be null: targetRange
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1313)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1280)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.NullPointerException: Property must not be null: targetRange
	at org.eclipse.lsp4j.util.Preconditions.checkNotNull(Preconditions.java:29)
	at org.eclipse.lsp4j.LocationLink.<init>(LocationLink.java:59)
	at org.eclipse.lsp4j.LocationLink.<init>(LocationLink.java:64)
	at com.redhat.qute.services.QuteDefinition.findDefinitionFromStartEndTagSection(QuteDefinition.java:175)
	at com.redhat.qute.services.QuteDefinition.findDefinitionFromSection(QuteDefinition.java:123)
	at com.redhat.qute.services.QuteDefinition.findDefinition(QuteDefinition.java:86)
	at com.redhat.qute.services.QuteLanguageService.findDefinition(QuteLanguageService.java:115)
	at com.redhat.qute.ls.template.TemplateFileTextDocumentService.lambda$definition$13(TemplateFileTextDocumentService.java:179)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1311)
	... 7 more
```

--- a/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/Section.java
+++ b/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/Section.java
@@ -30,6 +30,8 @@ public class Section extends Node implements ParametersContainer {
 
 	private final String tag;
 
+	private int startTagOpenOffset;
+
 	private int startTagCloseOffset;
 
 	private int endTagOpenOffset;
@@ -43,8 +45,9 @@ public class Section extends Node implements ParametersContainer {
 	public Section(String tag, int start, int end) {
 		super(start, end);
 		this.tag = tag;
-		this.endTagOpenOffset = NULL_VALUE;
+		this.startTagOpenOffset = NULL_VALUE;
 		this.startTagCloseOffset = NULL_VALUE;
+		this.endTagOpenOffset = NULL_VALUE;
 		this.endTagCloseOffset = NULL_VALUE;
 	}
 
@@ -66,7 +69,7 @@ public class Section extends Node implements ParametersContainer {
 	 * @return the offset which opens the section start tag.
 	 */
 	public int getStartTagOpenOffset() {
-		return getStart();
+		return startTagOpenOffset;
 	}
 
 	/**
@@ -81,6 +84,10 @@ public class Section extends Node implements ParametersContainer {
 	 */
 	public int getStartTagNameOpenOffset() {
 		return getStartTagOpenOffset() + 1;
+	}
+
+	void setStartTagOpenOffset(int startTagOpenOffset) {
+		this.startTagOpenOffset = startTagOpenOffset;
 	}
 
 	/**
@@ -141,19 +148,32 @@ public class Section extends Node implements ParametersContainer {
 	 *         #each) and false otherwise.
 	 */
 	public boolean isInStartTagName(int offset) {
+		if (!hasStartTag()) {
+			return false;
+		}
 		if (!isStartTagClosed()) {
 			// cases
 			// - {#|
 			// - {|#
 			return true;
 		}
-		if (offset > getStart() && offset <= getStartTagNameCloseOffset()) {
+		if (offset > getStartTagOpenOffset() && offset <= getStartTagNameCloseOffset()) {
 			// cases:
 			// - {#each| }
 			// - {|#each }
 			return true;
 		}
 		return false;
+	}
+
+	/**
+	 * Returns true if has a start tag.
+	 *
+	 *
+	 * @return true if has a start tag.
+	 */
+	public boolean hasStartTag() {
+		return getStartTagOpenOffset() != NULL_VALUE;
 	}
 
 	// ---------------------------- End tag methods
@@ -224,7 +244,7 @@ public class Section extends Node implements ParametersContainer {
 		if (!hasEndTag()) {
 			return false;
 		}
-		if (offset >= endTagOpenOffset + (afterBackSlash ? 1 : 0) && offset < getEnd()) {
+		if (offset > getEndTagOpenOffset() + (afterBackSlash ? 1 : 0) && offset < getEnd()) {
 			// cases:
 			// - {|/each}
 			// - {/|each}

--- a/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/TemplateParser.java
+++ b/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/TemplateParser.java
@@ -333,9 +333,10 @@ public class TemplateParser {
 			if (token == TokenType.StartTag) {
 				tag = scanner.getTokenText();
 			}
-			Section child = sectionFactory.createSection(tag, startSectionOffset, endSectionOffset);
-			curr.addChild(child);
-			curr = child;
+			Section section = sectionFactory.createSection(tag, startSectionOffset, endSectionOffset);
+			section.setStartTagOpenOffset(startSectionOffset);
+			curr.addChild(section);
+			curr = section;
 		}
 		return curr;
 	}

--- a/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDefinition.java
+++ b/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDefinition.java
@@ -137,16 +137,14 @@ class QuteDefinition {
 
 	private static boolean findDefinitionFromStartEndTagSection(int offset, Section section, Template template,
 			List<LocationLink> locations) {
-		Range originRange = null;
-		Range targetRange = null;
 		if (section.isInStartTagName(offset)) {
-			originRange = QutePositionUtility.selectStartTagName(section);
 
 			// 1. Jump to custom tag declared in the the {#insert custom-tag of the included
 			// Qute template (by using {#include base).
 			if (section.getSectionKind() == SectionKind.CUSTOM) {
 				QuteProject project = template.getProject();
 				if (project != null) {
+					Range originRange = null;
 					Node parent = section.getParent();
 					while (parent != null) {
 						if (parent.getKind() == NodeKind.Section) {
@@ -159,6 +157,9 @@ class QuteDefinition {
 									for (QuteIndex index : indexes) {
 										String linkedTemplateUri = index.getTemplatePath().toUri().toString();
 										Range linkedTargetRange = index.getRange();
+										if (originRange == null) {
+											originRange = QutePositionUtility.selectStartTagName(section);
+										}
 										locations.add(new LocationLink(linkedTemplateUri, linkedTargetRange,
 												linkedTargetRange, originRange));
 									}
@@ -170,16 +171,21 @@ class QuteDefinition {
 				}
 			}
 
-			// 2. Jump to end tag section
-			targetRange = QutePositionUtility.selectEndTagName(section);
-			locations.add(new LocationLink(template.getUri(), targetRange, targetRange, originRange));
+			if (section.hasEndTag()) {
+				// 2. Jump to end tag section
+				Range originRange = QutePositionUtility.selectStartTagName(section);
+				Range targetRange = QutePositionUtility.selectEndTagName(section);
+				locations.add(new LocationLink(template.getUri(), targetRange, targetRange, originRange));
+			}
 
 			return true;
 		} else if (section.isInEndTagName(offset)) {
-			// Jump to start tag section
-			originRange = QutePositionUtility.selectEndTagName(section);
-			targetRange = QutePositionUtility.selectStartTagName(section);
-			locations.add(new LocationLink(template.getUri(), targetRange, targetRange, originRange));
+			if (section.hasStartTag()) {
+				// Jump to start tag section
+				Range originRange = QutePositionUtility.selectEndTagName(section);
+				Range targetRange = QutePositionUtility.selectStartTagName(section);
+				locations.add(new LocationLink(template.getUri(), targetRange, targetRange, originRange));
+			}
 			return true;
 		}
 		return false;

--- a/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/definition/QuteDefinitionInIfSectionTest.java
+++ b/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/definition/QuteDefinitionInIfSectionTest.java
@@ -32,4 +32,11 @@ public class QuteDefinitionInIfSectionTest {
 		testDefinitionFor(template, "test.qute", //
 				ll("test.qute", r(1, 7, 1, 12), r(0, 6, 0, 11)));
 	}
+
+	@Test
+	public void undefinedVariable() throws Exception {
+		String template = "{#if val|ue}";
+		testDefinitionFor(template);
+	}
+
 }

--- a/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/definition/QuteDefinitionInSectionTagTest.java
+++ b/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/definition/QuteDefinitionInSectionTagTest.java
@@ -1,0 +1,99 @@
+/*******************************************************************************
+* Copyright (c) 2022 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package com.redhat.qute.services.definition;
+
+import static com.redhat.qute.QuteAssert.ll;
+import static com.redhat.qute.QuteAssert.r;
+import static com.redhat.qute.QuteAssert.testDefinitionFor;
+
+import org.junit.jupiter.api.Test;
+
+/**
+ * Test definition with section tag.
+ * 
+ * @author Angelo ZERR
+ *
+ */
+public class QuteDefinitionInSectionTagTest {
+
+	@Test
+	public void definitionInStartTagSection() throws Exception {
+		String template = "{|#each items}\r\n" + //
+				"		{it.name}\r\n" + //
+				"{/each}";
+		testDefinitionFor(template, "test.qute", //
+				ll("test.qute", r(0, 1, 0, 6), r(2, 1, 2, 6)));
+
+		template = "{#ea|ch items}\r\n" + //
+				"		{it.name}\r\n" + //
+				"{/each}";
+		testDefinitionFor(template, "test.qute", //
+				ll("test.qute", r(0, 1, 0, 6), r(2, 1, 2, 6)));
+
+		template = "{#each| items}\r\n" + //
+				"		{it.name}\r\n" + //
+				"{/each}";
+		testDefinitionFor(template, "test.qute", //
+				ll("test.qute", r(0, 1, 0, 6), r(2, 1, 2, 6)));
+
+		template = "{#each |items}\r\n" + //
+				"		{it.name}\r\n" + //
+				"{/each}";
+		testDefinitionFor(template);
+
+	}
+
+	@Test
+	public void definitionInEndTagSection() throws Exception {
+		String template = "{#each items}\r\n" + //
+				"		{it.name}\r\n" + //
+				"{|/each}";
+		testDefinitionFor(template, "test.qute", //
+				ll("test.qute", r(2, 1, 2, 6), r(0, 1, 0, 6)));
+
+		template = "{#each items}\r\n" + //
+				"		{it.name}\r\n" + //
+				"{/e|ach}";
+		testDefinitionFor(template, "test.qute", //
+				ll("test.qute", r(2, 1, 2, 6), r(0, 1, 0, 6)));
+
+		template = "{#each items}\r\n" + //
+				"		{it.name}\r\n" + //
+				"{/each|}";
+		testDefinitionFor(template, "test.qute", //
+				ll("test.qute", r(2, 1, 2, 6), r(0, 1, 0, 6)));
+
+		template = "{#each items}\r\n" + //
+				"		{it.name}\r\n" + //
+				"|{/each}";
+		testDefinitionFor(template);
+
+		template = "{#each items}\r\n" + //
+				"		{it.name}\r\n" + //
+				"{/each}|";
+		testDefinitionFor(template);
+
+	}
+
+	@Test
+	public void noStartTag() throws Exception {
+		String template = "{/|each}";
+		testDefinitionFor(template);
+	}
+
+	@Test
+	public void noEndTag() throws Exception {
+		String template = "{#|each items}";
+		testDefinitionFor(template);
+	}
+
+}

